### PR TITLE
[Feature] Log each entropy for composite distributions in PPO

### PR DESCRIPTION
**Pull request by louisfaury**
## Description

This PR enables PPO to log the entropy of each individual head of a composite policy separately.  

Concretely, for a composite distribution with, say, a nested discrete and continuous head, the `td_out` is augmented with some detached values.
```python
>>> loss = PPO(td)
>>> loss.keys()
[..., entropy, discrete-head-entropy, continuous-head-entropy, ...]
```
## Motivation and Context

This is an extremely useful debugging tool when training composite policies.

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: torchrl/objectives/ppo.py [hunk 1]
```diff
@@ -681,10 +679,17 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
             log_weight = log_weight.view(advantage.shape)
         neg_loss = log_weight.exp() * advantage
         td_out = TensorDict({"loss_objective": -neg_loss}, batch_size=[])
+        td_out.set("kl_approx", kl_approx.detach().mean())  # for logging
```
Comment (louisfaury) [inline torchrl/objectives/ppo.py:682]: Unrelated to the main topic of this MR, but afaik the existence and relevance of `kl_approx` is independent of whether we are training with an entropy bonus.

File: torchrl/objectives/ppo.py [hunk 2]
```diff
@@ -681,10 +679,17 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
         if self.entropy_bonus:
-            entropy = self.get_entropy_bonus(dist)
+            entropy = self._get_entropy(dist)
+            if is_tensor_collection(entropy):
+                # Logs the entropy of each action head.
+                for head_key, head_entropy in entropy.items(
+                    include_nested=True, leaves_only=True
+                ):
+                    td_out.set("-".join(head_key), head_entropy.detach().mean())
```
Comment (louisfaury) [inline torchrl/objectives/ppo.py:690]: Choosing under which key to log the individual factor's entropy was a bit of a headache. The way I personally use `CompositeDistribution` yields tensordict that look like:

```
action: {
    head_1: {
        action: ...
        entropy: ...
   }
   head_2 {
        action: ...
        entropy: ...
   }
}
```
which means that using `head_key[-1]` to log each entropy is not really a viable solution (all the factor entropies will be logged under the same name, `entropy`). I'm not sure how to get a one-size-fits-all here, and happy for suggestions. The current solution ensures that there is no collision, at the price of having very verbose keys.

Comment (vmoens) [reply torchrl/objectives/ppo.py:690]: Wouldn't the most generic solution just be to log the entropy TD as it comes?
Why do we need to rename it?
BTW it seems to me that what you're doing here amends to
```python
tensordict.flatten_keys("-").detact().mean()
```
Nit: this isn't collision-safe I think (but flatten_keys will tell you if there are any collision):
eg `("key-one", "entropy")` and `("key", "one", "entropy")` will collide

File: torchrl/objectives/ppo.py [hunk 3]
```diff
@@ -976,11 +981,18 @@ def forward(self, tensordict: TensorDictBase) -> TensorDictBase:
...
 
         if self.entropy_bonus:
-            entropy = self.get_entropy_bonus(dist)
+            entropy = self._get_entropy(dist)
+            if is_tensor_collection(entropy):
+                # Logs the entropy of each action head.
+                for head_key, head_entropy in entropy.items(
+                    include_nested=True, leaves_only=True
+                ):
+                    td_out.set("-".join(head_key), head_entropy.detach().mean())
```
Comment (vmoens) [inline torchrl/objectives/ppo.py:993]: ditto, we could not flatten and let users do that, no?

File: torchrl/objectives/ppo.py (hunk 2 above)
Comment (vmoens) [reply torchrl/objectives/ppo.py:690]: also - are we 100% sure all keys are nested? I think so (that's how `CompositeDist` works) but maybe we could just put a safeguard check here to make sure an error is raised if that assumption is violated (eg, users have their own dist class that returns `{"entropy", ("nested", "entropy")}` keys).

**Review by vmoens — COMMENTED** [review]
LGTM, just a couple of comments to address before we merge it

LMK what you think would be the best way to log the entropies in the output data structure, I think flattening may be a bit surprising

File: torchrl/objectives/ppo.py (hunk 2 above)
Comment (louisfaury) [reply torchrl/objectives/ppo.py:690]: I followed your recommendation: I added a `composite_entropy` key to the loss td. Two remarks:

1.  The composite entropy is not logged under `entropy` to avoid BC (users currently expect a Tensor),
2. I did not `detach()` the composite entropy; this would allow the user to compute a custom entropy bonus when using a composite entropy (e.g. not the same penalty per head).

Wdyt ?

Comment (vmoens) [reply torchrl/objectives/ppo.py]: Not detaching could be slightly bc-breaking: what if I do `loss_tensordict.sum(reduce=True).backward()`?
Previously, this was giving the right result, now it would also backprop through the entropy. Usually metadata in loss outputs is guaranteed to be non-differentiable, that would be a one off.
But I understand that it could be useful... 
We could add a kwarg in the constructor (which would become a bit overloaded!)

Comment (louisfaury) [reply torchrl/objectives/ppo.py]: Right. I'll detach for now, let's revisit when there is a need/ask for it?

Comment (vmoens) [reply torchrl/objectives/ppo.py]: yes, thanks for raising it!
Thinking about it, a way of doing this could be to pass the entropy coefficient as a tensordict and do 
`(td_coef * td_entropy).sum(reduce=True)`
Idea for a follow-up PR ;)

**Review by vmoens — APPROVED** [review]
LGTM thanks!
